packages/core/database/lib/entity-manager.js

```diff
@@ -615,40 +627,117 @@ const createEntityManager = (db) => {
-          ) {
+          if (onlyDeleteRelation) {
+            // clear previous associations in the joinTable
             await this.createQueryBuilder(joinTable.name)
               .delete()
-              .where({ [inverseJoinColumn.name]: toIds(data[attributeName]) })
+              .where({ [joinColumn.name]: id })
               .where(joinTable.on || {})
               .execute();
-          }
+          } else {
+            const cleanRelationData = toAssocs(data[attributeName]);
+            const connect = cleanRelationData.connect || [];
+            const disconnect = cleanRelationData.disconnect || [];
+            const isPartialUpdate = connect.length || disconnect.length;
+
+            if (isPartialUpdate) {
+              const idsToRemove = _.map('id', connect.concat(disconnect));
+              const existingRelsToDelete = await this.createQueryBuilder(joinTable.name)
+                .select([inverseJoinColumn.name, 'order'])
+                .where({
+                  [joinColumn.name]: id,
+                  [inverseJoinColumn.name]: { $in: idsToRemove },
+                })
+                .where(joinTable.on || {})
+                .execute();
 
-          if (!_.isNull(data[attributeName])) {
-            const insert = toAssocs(data[attributeName]).map((data) => {
-              return {
-                [joinColumn.name]: id,
-                [inverseJoinColumn.name]: data.id,
-                ...(joinTable.on || {}),
-                ...(data.__pivot || {}),
-              };
-            });
+              await this.createQueryBuilder(joinTable.name)
+                .delete()
+                .where({
+                  [joinColumn.name]: id,
+                  [inverseJoinColumn.name]: { $in: idsToRemove },
+                })
+                .where(joinTable.on || {})
+                .execute();
 
-            // if there is nothing to insert
-            if (insert.length === 0) {
-              continue;
-            }
+              for (const relToDelete of existingRelsToDelete) {
+                await this.createQueryBuilder(joinTable.name)
+                  .update({ order: db.getConnection().raw('?? - 1', 'order') })
+                  .where({
+                    [joinColumn.name]: id,
+                    order: { $gt: relToDelete.order },
+                  })
+                  .where(joinTable.on || {})
+                  .execute();
+              }
+
+              if (connect.length) {
+                const { max } = await this.createQueryBuilder(joinTable.name)
+                  .max('order')
+                  .where({ [joinColumn.name]: id })
+                  .where(joinTable.on || {})
+                  .first()
+                  .execute();
+
+                insert = _.uniqBy('id', connect).map((rel, idx) => ({
+                  [joinColumn.name]: id,
+                  [inverseJoinColumn.name]: rel.id,
+                  order: max + idx + 1,
```

Comment:
Yes I think, otherwise how do you keep the order stable?

Reply:
If you are only adding new relations at the end , the order will always be stable I think

Reply:
It depends on the database actually, it tends to be sorted by creation date but there is no guarantee to it.
What we could do is to sort by id (since the id is incremental) that would work. However since we need to keep the order from the 2 sides, we need the 2 order columns and the id would not be enough then. So we almost could have avoided the order columns yes :/
(the 2 columns are not implemented yet)

Reply:
True, did not think about the order could be different in the database!